core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java

```diff
@@ -197,16 +197,41 @@ public int unfinishedRequests() {
     }
 
     @Override
-    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
-        super.channelUnregistered(ctx);
-        unfinishedResponses.keySet().forEach(StreamMessage::abort);
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        // Give the unfinished streaming responses a chance to close themselves before we abort them,
+        // so that successful responses are not aborted due to a race condition like the following:
+        //
+        // 1) A publisher of a response stream sends the complete response
+        //    but does not call StreamWriter.close() just yet.
+        // 2) An HTTP/1 client receives the complete response and closes the connection, which is totally fine.
+        // 3) The response stream is aborted once the server detects the disconnection.
+        // 4) The publisher calls StreamWriter.close() but it's aborted already.
+        //
+        // To reduce the chance of such situation, we wait a little bit before aborting unfinished responses.
+
+        switch (protocol) {
+            case H1C:
+            case H1:
+                // XXX(trustin): How much time is 'a little bit'?
+                ctx.channel().eventLoop().schedule(this::cleanup, 1, TimeUnit.SECONDS);
```

Comment:
I'm not sure if we should make this delay configurable.

Reply:
Configuring just this knob is probably overkill - setting it based on idle timeout handler might make sense, but might also be overkill.